src/ferny/interaction_agent.py

```diff
@@ -251,6 +251,15 @@ async def communicate(self) -> None:
             try:
                 # We handle fds very carefully to avoid leaking them, even in case of exceptions
                 data, fds, _flags, _addr = recv_fds(self.ours, 4096, 10)
+                # read the rest of the request
+                while True:
+                    try:
+                        block = self.ours.recv(4096, socket.MSG_DONTWAIT)
```

Comment:
I had the impression (from testing) that the fds would always arrive with the last byte of the message that they were sent with.  Maybe that's not the case?

I'd like to test this very thoroughly to make sure we know exactly what the ordering guarantee is here.  As its written, this loop could easily pick up data from the following request.

Reply:
A simpler approach would of course just be to increase the 4096 size of the first read to something really large, maybe 65536? But that would just make the bug less likely to happen, not really fix it.

My immediate selfish reason was "I want `pytest` to succeed", but for production quality we should support reading the request in multiple chunks IMHO.

But good point about the fds being in the last chunk! The test case seems to work (albeit with the random hang, see above), and locally `pytest` is always happy, but of couse we can add more corner cases.

Reply:
We actually *do* support reading multiple chunks.  We append `data` to `self.buffer` and only once we receive the `fds` then we do the interaction based on the accumulated `self.buffer`.

Reply:
Then demonstrably the fds get sent with the first block, not the last one. Are you saying that `pytest` works for you with a large environment? (I'm not actually how exactly `$LS_COLORS` gets into my environment, but it should be easy enough to define a large env variable -- that's exactly what the unit test does)

Reply:
Nope.  It absolutely fails for me too.

Reply:
I did a test.  It seems that, more or less, the file descriptors end up associated with the first byte of the data that they were sent with.  Whenever that byte gets read off the wire, you get your fds.

It's tricky, though: the kernel will never combine lists of fds (ie: reading multiple small fd-containing messages will effectively get you packet semantics) but it's happy to combine non-fd-containing messages that came before.

Ie: if you:

  - send 'abc'
  - send '123' + fd

then the kernel will happily give you 'abc123' + fd.

This is a headache...